docs/develop/dotnet/message-passing.mdx

```diff
@@ -1,207 +1,686 @@
+
+Here are two important use cases for `Workflow.await`:
 
-**How to define a Query using the Temporal .NET SDK**
+- Waiting in a handler until it is appropriate to continue.
+- Waiting in the main Workflow until all active handlers have finished.
 
-A Query has a name and can have arguments.
+The condition state you're waiting for can be updated by and reflect any part of the Workflow code.
+This includes the main Workflow method, other handlers, or child coroutines spawned by the main Workflow method, and so forth.
 
-- The name, also called a Query type, is a string.
-- The arguments must be [serializable](/dataconversion).
+#### Use wait conditions in handlers {#wait-in-handlers}
 
-Queries may be methods or properties (only the getter is used).
-Queries must be synchronous and must not mutate workflow state in any way or issue any [Commands](/workflows#command).
-To define a Query, set the `[WorkflowQuery]`(https://dotnet.temporal.io/api/Temporalio.Workflows.WorkflowQueryAttribute.html) attribute on the Query method or property inside your Workflow.
-A string name can be provided to the attribute, otherwise the Query's name defaults to the unqualified method/property name.
+Sometimes, async Signal or Update handlers need to meet certain conditions before they should continue.
+Using a wait condition with [`Workflow.WaitConditionAsync`](https://dotnet.temporal.io/api/Temporalio.Workflows.Workflow.html?#Temporalio_Workflows_Workflow_WaitConditionAsync_System_Func_System_Boolean__System_Int32_System_Nullable_System_Threading_CancellationToken__) sets a function that prevents the code from proceeding until the condition returns `true`.
+This is an important feature that helps you control your handler logic.
 
-Queries can be methods that can accept arguments:
+Consider a `ReadyForUpdateToExecute` method that runs before your Update handler executes.
+The [`Workflow.WaitConditionAsync`](https://dotnet.temporal.io/api/Temporalio.Workflows.Workflow.html?#Temporalio_Workflows_Workflow_WaitConditionAsync_System_Func_System_Boolean__System_Int32_System_Nullable_System_Threading_CancellationToken__) call waits until your condition is met:
 
 ```csharp
-[WorkflowQuery]
-public string GetMyStatus(MyStatusParam input) => statuses[input.Type];
+    [WorkflowUpdate]
+    public async Task<string> MyUpdateAsync(UpdateInput updateInput)
+    {
+        await Workflow.WaitConditionAsync(() => ReadyForUpdateToExecute(updateInput));
+        // ...
+    }
 ```
 
-Or properties:
+Remember: Handlers can execute before the main Workflow method starts.
+
+#### Ensure your handlers finish before the Workflow completes {#wait-for-message-handlers}
+
+Workflow wait conditions can ensure your handler completes before a Workflow finishes.
+When your Workflow uses async Signal or Update handlers, your main Workflow method can return or continue-as-new while a handler is still waiting on an async task, such as an Activity result.
```

Comment:
```suggestion
When your Workflow uses async Signal or Update handlers, your main Workflow method can return or continue-as-new while a handler is still waiting on an async Task, such as an Activity result.
```

Reply:
I don't think "task" should always be upper-cased. Perhaps in some contexts when it's very explicitly an abbreviation for Activity Task or Workflow Task, but here we're referring to tasks in the .NET's concurrency framework (which we do in several languages, e.g. Typescript (Node event loop tasks) and Python (asyncio tasks.)